callbacks/opentelemetry.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// HandleChainStart implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleChainStart(_ context.Context, _ map[string]any) {
+}
+
+// HandleLLM implements Handler.
+func (o *OpenTelemetryCallbacksHandler) HandleLLM(ctx context.Context, messages []llms.MessageContent, options llms.CallOptions, next func(ctx context.Context) (*llms.ContentResponse, error)) (*llms.ContentResponse, error) {
+	ctx, span := o.tracer.Start(ctx, "gen_ai.model."+options.Model+".completion")
+	defer span.End()
+	response, err := next(ctx)
+	if err != nil {
+		span.RecordError(err)
+	}
+
+	o.processLLMRequest(span, messages, options)
+	o.processLLMResponse(span, response, options)
+
+	return response, err
+}
+
+// processLLMRequest records information about the request to the span.
+func (o *OpenTelemetryCallbacksHandler) processLLMRequest(span trace.Span, messages []llms.MessageContent, options llms.CallOptions) {
+	o.recordRequestAttributes(options, span)
+
+	if o.opts.logPrompts {
+		o.logPrompts(messages, span)
+	}
+}
+
+// recordRequestAttributes records the request attributes to the span.
+func (*OpenTelemetryCallbacksHandler) recordRequestAttributes(options llms.CallOptions, childSpan trace.Span) {
+	llmModeValue := "text"
+	if options.JSONMode {
+		llmModeValue = "json"
+	}
+
+	requestAttributes := []attribute.KeyValue{
+		llmModel.String(options.Model),
+		llmVendor.String(options.Vendor),
+		llmMode.String(llmModeValue),
+		llmTemperature.Float64(float64(options.Temperature)),
+		llmTopP.Float64(float64(options.TopP)),
+		llmTopK.Float64(float64(options.TopK)),
+		llmStopSequences.StringSlice(options.StopWords),
+		llmFrequencyPenalty.Float64(options.FrequencyPenalty),
+		llmPresencePenalty.Float64(float64(options.PresencePenalty)),
+		llmMaxTokens.Int(options.MaxTokens),
+	}
+	childSpan.SetAttributes(requestAttributes...)
+}
+
+// logPrompts logs the prompts to the span as events.
+func (o *OpenTelemetryCallbacksHandler) logPrompts(messages []llms.MessageContent, span trace.Span) {
+	requestInfoMessages := o.transformPromptMessages(messages)
+
+	for _, message := range requestInfoMessages {
+		span.AddEvent("gen_ai.prompt", trace.WithAttributes(
+			llmMessageIndex.Int(message.Index),
+			llmMessageRole.String(message.Role),
+			llmMessageContent.String(message.Content),
+		))
+	}
+}
+
+// transformPromptMessages transforms the Langchain-Go messages to a format that can be logged as events in the span, in accordance with the semantic conventions for GenAI LLM spans.
+func (o *OpenTelemetryCallbacksHandler) transformPromptMessages(messages []llms.MessageContent) []semconvGenAIMessage {
+	semconvGenAIMessages := []semconvGenAIMessage{}
+	for idx, message := range messages {
+		msg := &semconvGenAIMessage{
+			Index:   idx,
+			Role:    string(message.Role),
+			Content: "",
+		}
+		for _, part := range message.Parts {
+			switch part := part.(type) {
+			case llms.TextContent:
+				msg.Content += part.String()
+			case llms.ToolCall:
```

Comment:
I would suggest for first PR to not handle tools, etc. as there is an open discussion on that right now. We can do it in a follow-up? that's also less unit tests etc for first PR. Basically start with chat, then skip others until a PR which elaborates and tests them. wdyt?